staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -353,8 +374,31 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
 	a.cachedCreds = newCreds
 	// Only close all connections when TLS cert rotates. Token rotation doesn't
 	// need the extra noise.
-	if a.onRotate != nil && oldCreds != nil && !reflect.DeepEqual(oldCreds.cert, a.cachedCreds.cert) {
-		a.onRotate()
+	if oldCreds != nil && !reflect.DeepEqual(oldCreds.cert, a.cachedCreds.cert) {
+		// Can be nil if the exec auth plugin only returned token auth.
+		if oldCreds.cert != nil && oldCreds.cert.Leaf != nil {
+			metrics.ClientCertRotationAge.Observe(time.Now().Sub(oldCreds.cert.Leaf.NotBefore))
+		}
+		if a.onRotate != nil {
+			a.onRotate()
+		}
+	}
+	// Update the nearest client cert expiration from cache as it likely changed.
+	metrics.ClientCertExpiration.Set(globalCache.oldestExpiration())
+	return nil
+}
+
+// Gets the leaf certificate if cached certificate exists and is valid.
+// otherwise returns nil.
+func (a *Authenticator) leafCertificate() *x509.Certificate {
+	if a.cachedCreds != nil && a.cachedCreds.cert != nil {
+		if a.cachedCreds.cert.Leaf != nil {
+			return a.cachedCreds.cert.Leaf
+		} else if c, err := x509.ParseCertificate(a.cachedCreds.cert.Certificate[0]); err == nil {
```

Comment:
shouldn't we do this above after calling X509KeyPair to populate Leaf once per cert, rather than once per invocation of leafCertificate()?

Reply:
I am not sure honestly. The X509 library doesn't doesn't really define when cert.Leaf should be populated. I would like to populate Leaf but I have a few issues with it from a code cleanliness perspective.

1. Its a modifying operation in a function that isn't super clear it modifies the Authorization object.
2. It breaks test (that I can fix). Because the actual x509.Certificate being compared in got/want changes if leafCertificate() is called.

Thoughts?

Reply:
Fixed,